pageserver/src/tenant.rs

```diff
@@ -1389,13 +1445,107 @@ impl Tenant {
+                let rx = Arc::new(rx);
+                let this = self.clone();
+                let timeline = timeline.clone();
+
+                // TODO: this could be tenant scoped task_mgr task?
+                tokio::spawn({
+                    let rx = rx.clone();
+                    async move {
+                        // to uphold the MaybeDone promise, we keep the channel alive *until* we've
+                        // swapped the MaybeDone::Pending -> MaybeDone::Done value
+                        let tx = tx;
+                        let rx = rx;
+
+                        let res = this.unique_delete_timeline(&timeline).await;
+
+                        {
+                            let mut g = timeline.delete_self.lock().await;
+                            #[cfg(debug_assertions)]
+                            match g.as_ref() {
+                                Some(timeline::MaybeDone::Pending(weak)) => {
+                                    let same = weak
+                                        .upgrade()
+                                        // we don't yet have Receiver::same_channel
+                                        .map(|rx2| Arc::ptr_eq(&rx, &rx2))
+                                        .unwrap_or(false);
+                                    assert!(same, "different channel had been replaced or dropped");
+                                }
+                                other => panic!("unexpected MaybeDone: {other:?}"),
+                            }
+                            *g = Some(timeline::MaybeDone::Done(res.clone()));
+                            drop(rx);
+                        }
+
+                        // now no one can get the Pending(weak) value to upgrade and they only see
+                        // the Done(res).
+
+                        // send the result value to listeners, if any
+                        drop(tx.send(res));
+                    }
+                    .instrument(tracing::info_span!("unique_delete_timeline"))
+                });
+                *g = Some(timeline::MaybeDone::Pending(Arc::downgrade(&rx)));
+                rx
+            }
+            // subscribe while the mutex is held, so we are guaranteed to receive the message,
+            // unless a panic comes.
+            .resubscribe()
+        };
+
+        {
+            use tokio::sync::broadcast::error::RecvError;
+
+            match rx.recv().await {
+                Ok(Ok(())) => Ok(()),
+                Ok(Err(e)) => Err(DeleteTimelineError::from(&e)),
+                // lagged doesn't mean anything with 1 send, but whatever, handle it the same
+                Err(RecvError::Closed | RecvError::Lagged(_)) => {
+                    Err(DeleteTimelineError::InternalFailure)
+                }
+            }
+        }
+    }
+
+    async fn unique_delete_timeline(
+        &self,
+        timeline: &Arc<Timeline>,
+    ) -> Result<(), InnerDeleteTimelineError> {
```

Comment:
naming suggestion:

1. `fn delete_timeline_locked` or `fn delete_timeline_exclusive` or `fn delete_timeline_task`
2. The error type should be named correspondingly:
   - DeleteTimelineLockedError
   - DeleteTimelineExclusiveError
   - DeleteTimelineTaskError